Fix the carryall freeze and untangle the afterLandActivity mess

abcdefg30 opened the pull request:
Fixes #14572.
Supersedes #14595.

This PR removes `MovingToResources`, `MovingToRefinery` and `MovementCancelled` from the `INotifyHarvesterAction` interface, as those methods are redundant to what `ICallForTransport` offers.
We also avoid getting stuck in a loop by not creating a new activity and queuing another new one as `landAfterActivity` right afterwards.

abcdefg30 committed:
Introduce a new CarryableHarvester trait

abcdefg30 commented:
Should be fixed now.

GraionDilach commented:
This literally breaks the RA2 Chrono harvester.

abcdefg30 commented:
Hm, one could work around that by using the `ICallForTransport` interface, that's not very nice though. On a quick thought: What about having `Harvester` + its activities only care about `INotifyHarvesterAction` (without removing the extra interface methods) and having `AutoCarryable` only care about `ICallForTransport` (fulfilling the second request behind this PR) with a third trait bridging harvester actions and transport requests?

reaperrr commented:
> with a third trait bridging harvester actions and transport requests?

What about a `CallsForTransport` bool flag on `Harvester` instead?

Your proposal might be more sensible, right now I just can't picture how that bridging trait would look like (let alone how to call it).

Unit158 commented:
This should be labeled "Unaddressed comments" not "Unadressed comments"

pchote commented:
@GraionDilach's point about the chrono harvester invalidates the original motivation for removing `INotifyHarvesterAction.MovingToRefinery` etc.  Considering that, keeping `AutoCarryable : INotifyHarvesterAction` is IMO the most sensible implementation.  Can we revert that change from here?

GraionDilach commented:
In that case, I would actually also file an issue about axing the interfaces, instantly assigned to the Future milestone with the flashed out locomotor setup being it's prerequisite - technically, the RA2 harvester could jump to any dock (service depot included) and it would make sense to shift the implementation to a locomotor longterm instead of what it is atm.

pchote commented:
Ping @abcdefg30.  It would be good to get some progress here both to fix this crash finally for the next release, and also to help clean up the PR queue :smile:

pchote commented:
If we can get this ready and reviewed in time for the playtest, then great.  I don't think we can justify delaying the playtest much more specifically for this, however.  Bumping to Next + 1.

abcdefg30 committed:
Further untangle (and - hopefully - fix) the afterLandActivity mess

abcdefg30 commented:
Rebased & updated.

Added a new `CarryableHarvester` trait that does the bridging between `INotifyHarvesterAction` and `ICallForTransport`. (That was previously done in `AutoCarryable`, so that trait is "untangled" now as well.)

GraionDilach commented:
This will need an upgrade rule?

abcdefg30 commented:
Yes, that might be a good idea.

abcdefg30 commented:
Rebased and added a rule.

pchote commented:
Needs a rebase, and the update rule should move to the new directory.

abcdefg30 commented:
Rebased.

reaperrr commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -233,11 +233,8 @@ public void UnblockRefinery(Actor self)
-					var findResources = new FindResources(self);
-					foreach (var n in notify)
-						n.MovingToResources(self, moveTo, findResources);
+					// FindResources cares about calling INotifyHarvesterAction
```
'cares about' -> 'takes care of'

reaperrr commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -381,12 +378,10 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 
-				foreach (var n in notify)
-					n.MovingToResources(self, loc.Value, findResources);
+				// FindResources cares about calling INotifyHarvesterAction
```
'cares about' -> 'takes care of'

reaperrr commented:
Otherwise looks good to me, needs a rebase as well, though.

abcdefg30 committed:
Add an update rule for CarryableHarvester

abcdefg30 commented:
Updated.

pchote commented:
I have pushed an update to this branch:
 * Fix a typo in the second commit's message
 * Rebase
 * Fix the update rule to not refer to a specific line number

pchote reviewed (approved):
Spectated a very long D2K AI game without any noticable issues. Code LGTM.
